crates/guest-rust/rt/src/async_support/stream_support.rs

```diff
@@ -1,133 +1,187 @@
+    }
+
+    /// Same as [`StreamWriter::write`], except this takes [`AbiBuffer<T>`]
+    /// instead of `Vec<T>`.
+    pub fn write_buf(&mut self, values: AbiBuffer<T>) -> StreamWrite<'_, T> {
+        StreamWrite {
+            op: WaitableOperation::new((self, values)),
         }
     }
 
-    /// Cancel the current pending write operation.
+    /// Writes all of the `values` provided into this stream.
     ///
-    /// This will panic if no such operation is pending.
-    pub fn cancel(&mut self) {
-        assert!(self.future.is_some());
-        self.future = None;
+    /// This is a higher-level method than [`StreamWriter::write`] and does not
+    /// expose cancellation for example. This will successively attempt to write
+    /// all of `values` provided into this stream. Upon completion the same
+    /// vector will be returned and any remaining elements in the vector were
+    /// not sent because the stream was closed.
+    pub async fn write_all(&mut self, values: Vec<T>) -> Vec<T> {
+        // Perform an initial write which converts `values` into `AbiBuffer`.
+        let (mut status, mut buf) = self.write(values).await;
+
+        // While the previous write completed and there's still remaining items
+        // in the buffer, perform another write.
+        while let StreamResult::Complete(_) = status {
+            if buf.remaining() == 0 {
+                break;
+            }
+            (status, buf) = self.write_buf(buf).await;
+        }
+
+        // Return back any values that weren't written by shifting them to the
+        // front of the returned vector.
+        assert!(buf.remaining() == 0 || matches!(status, StreamResult::Closed));
+        buf.into_vec()
     }
 
-    /// Close the writer with an error that will be returned as the last value
+    /// Writes the singular `value` provided
     ///
-    /// Note that this error is not sent immediately, but only when the
-    /// writer closes, which is normally a result of a `drop()`
-    pub fn close_with_error(self, err: ErrorContext) {
-        super::with_entry(self.handle, move |entry| match entry {
-            Entry::Vacant(_) => unreachable!(),
-            Entry::Occupied(mut entry) => match entry.get_mut() {
-                _ => {
-                    // Note: the impending drop after this function runs should trigger
-                    entry.insert(Handle::WriteClosedErr(Some(err)));
-                }
-            },
-        });
+    /// This is a higher-level method than [`StreamWriter::write`] and does not
+    /// expose cancellation for example. This will attempt to send `value` on
+    /// this stream.
+    ///
+    /// If the other end hangs up then the value is returned back as
+    /// `Some(value)`, otherwise `None` is returned indicating the value was
+    /// sent.
+    pub async fn write_one(&mut self, value: T) -> Option<T> {
+        // TODO: can probably be a bit more efficient about this and avoid
```

Comment:
FWIW, [this is how I addressed it on the host side](https://github.com/bytecodealliance/wasip3-prototyping/blob/80a85e686a1722469891d07c4246dc26f98dc322/crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs#L242-L262), i.e. by making the buffer type generic and implementing it for `Vec`, `Bytes`, and `Single`.  That would require significant refactoring here, though, given the way `AbiBuffer` is currently tied to `Vec`.

Reply:
Good point yeah, I kind of want to avoid the complexity for now but I  do still want to leave a note about a potential future optimization, so I'm probably going to leave this as is (and might advocate for the host to do the same in the near term...)

Reply:
To be clear, the host optimization was primarily motivated by being able to support `bytes::Bytes` (which is what Hyper uses out of the box, and thus `wasmtime-wasi-http`) natively without any conversions.  The single-element optimization was just a nice side-effect of generic-izing the code.